src/v/cluster/plugin_table.cc

```diff
@@ -0,0 +1,263 @@
+          "inconsistency: missing transform output index entry: {} id: {}",
+          output_topic,
+          id);
+        if (output_it->second.empty()) {
+            _output_topic_index.erase(output_it);
+        }
+    }
+    _name_index.erase(name_it);
+    _underlying.erase(it);
+    run_callbacks(id);
+}
+void plugin_table::reset_transforms(plugin_table::map_t snap) {
+    name_index_t snap_name_index;
+    topic_index_t snap_input_index;
+    topic_index_t snap_output_index;
+
+    // Perform a map diff to figure out which transforms need change callbacks
+    // fired.
+    std::vector<transform_id> all_deletes;
+    std::vector<transform_id> all_changed;
+    std::vector<transform_id> all_inserted;
+    for (const auto& [k, v] : _underlying) {
+        auto it = snap.find(k);
+        if (it == snap.end()) {
+            all_deletes.push_back(k);
+        } else if (v != it->second) {
+            all_changed.push_back(k);
+        }
+        // Otherwise unchanged
+    }
+    for (const auto& [k, v] : snap) {
+        if (!_underlying.contains(k)) {
+            all_inserted.push_back(k);
+        }
+        // build the name index
+        auto it = snap_name_index.insert({v.name, k});
+        if (!it.second) {
+            throw std::logic_error(ss::format(
+              "transform meta id={} is attempting to use a name {} which is "
+              "already registered to {}",
+              k,
+              v.name,
+              it.first->first));
+        }
+        // build topic indexes
+        snap_input_index[v.input_topic].emplace(k);
+        for (const auto& output_topic : v.output_topics) {
+            snap_output_index[output_topic].emplace(k);
+        }
+    }
+
+    // Do the actual swap
+    _underlying = std::move(snap);
+    _name_index = std::move(snap_name_index);
+    _input_topic_index = std::move(snap_input_index);
+    _output_topic_index = std::move(snap_output_index);
+
+    for (transform_id deleted : all_deletes) {
+        run_callbacks(deleted);
+    }
+    for (transform_id updated : all_changed) {
+        run_callbacks(updated);
+    }
+    for (transform_id inserted : all_inserted) {
+        run_callbacks(inserted);
+    }
+}
+plugin_table::notification_id
+plugin_table::register_for_updates(plugin_table::notification_callback cb) {
+    auto it = _callbacks.insert({++_latest_id, std::move(cb)});
+    vassert(it.second, "invalid duplicate in callbacks");
+    return _latest_id;
+}
+
+void plugin_table::unregister_for_updates(notification_id id) {
+    _callbacks.erase(id);
+}
+void plugin_table::run_callbacks(transform_id id) {
+    for (const auto& [_, cb] : _callbacks) {
+        cb(id);
+    }
+}
+bool plugin_table::name_less_cmp::operator()(
+  const model::transform_name& lhs, const std::string_view& rhs) const {
+    return lhs() < rhs;
+}
+bool plugin_table::name_less_cmp::operator()(
+  const std::string_view& lhs, const model::transform_name& rhs) const {
+    return lhs < rhs();
+}
```

Comment:
i think that if you do a default friend in c++20 that the left/right variants will be auto generated. not sure it works in all cases, but if does here you could remove some of this boilerplate.

Reply:
I could not get this to work on the model class, and it doesn't help that we need a comparator struct here. I will happily take suggestions here, but I couldn't get this to work right.